File: qap_cpu/src/PMX.cpp
#include "PMX.h"
#include <exception>
#include <algorithm>
#include <list>
#include <string>
#include <numeric>

Permutation<unsigned>
PMX::perform(const Permutation<unsigned> &first, const Permutation<unsigned> &second, unsigned start, unsigned end) {
    if (first.size() != second.size())
        throw std::invalid_argument{"Permutations have different sizes"};

    if (start >= first.size())
        throw std::invalid_argument{"Invalid start index"};

    if (end >= first.size())
        throw std::invalid_argument{"Invalid end index"};


    Permutation<unsigned> result(first.size(), std::numeric_limits<unsigned>::max());

    std::copy(first.begin() + start, first.begin() + end + 1, result.begin() + start);

    std::list<unsigned> tmp;

    for (unsigned i = start; i <= end; i++) {
        if (!isValueInPermutationSegment(second[i], first, start, end))
            tmp.push_back(second[i]);
    }

    for (unsigned value : tmp) {
        switchValue(value, value, first, second, result, start, end);
    }

    for (unsigned i = 0; i < result.size(); i++) {
        if (result[i] == std::numeric_limits<unsigned>::max())
            result[i] = second[i];
    }

    return result;
}

unsigned
PMX::getIndexOfValue(const Permutation<unsigned> &permutation, unsigned value) {
    for (unsigned i = 0; i < permutation.size(); i++) {
        if (permutation[i] == value)
            return i;
    }

    throw std::runtime_error{"Could not find value: " + std::to_string(value)};
}

bool
PMX::isValueInPermutationSegment(unsigned value, const Permutation<unsigned> &p, unsigned start, unsigned end) {
    return std::find(p.begin() + start, p.begin() + end + 1, value) != (p.begin() + end + 1);
}

void
PMX::switchValue(unsigned valueToSwitch, unsigned valueToWrite, const Permutation<unsigned> &firstParent, const Permutation<unsigned> &secondParent,
                 Permutation<unsigned> &child, unsigned start, unsigned end) {
    auto indexInSecondParent = getIndexOfValue(secondParent, valueToSwitch);
    auto valueFromFirstParent = firstParent[indexInSecondParent];
    indexInSecondParent = getIndexOfValue(secondParent, valueFromFirstParent);
    if (indexInSecondParent >= start && indexInSecondParent <= end) {
        switchValue(valueFromFirstParent, valueToWrite, firstParent, secondParent, child, start, end);
    } else {
        child[indexInSecondParent] = valueToWrite;
    }
}

File: em_alg/em_discrete/pmx.py
from typing import List


def pmx(first_permutation, second_permutation, lower, upper):
    if len(first_permutation) != len(second_permutation):
        raise RuntimeError("pmx: permutations have different size")

    if not 0 <= lower < len(first_permutation):
        raise RuntimeError("pmx: lower bound is too big ", lower)

    if not 0 <= upper < len(first_permutation):
        raise RuntimeError("pmx: upper bound is too big ", upper)

    if lower > upper:
        raise RuntimeError("pmx: lower bound is greater than upper")

    result = [-1] * len(first_permutation)
    result[lower:upper+1] = first_permutation[lower:upper+1]

    tmp = []
    for i in range(lower, upper + 1):
        if not second_permutation[i] in first_permutation[lower:upper+1]:
            tmp.append(second_permutation[i])

    for value in tmp:
        switch_value(value, value, first_permutation, second_permutation, result, lower, upper)

    for i in range(len(first_permutation)):
        if result[i] == -1:
            result[i] = second_permutation[i]

    return result


def switch_value(value_to_switch: int, value_to_write: int, first_permutation: List[int], second_permutation: List[int],
                 result_permutation: List[int], lower: int, upper: int):
    index_in_second_parent = second_permutation.index(value_to_switch)
    value_from_first_parent = first_permutation[index_in_second_parent]
    index_in_second_parent = second_permutation.index(value_from_first_parent)

    if lower <= index_in_second_parent <= upper:
        switch_value(value_from_first_parent, value_to_write, first_permutation, second_permutation, result_permutation,
                     lower, upper)
    else:
        result_permutation[index_in_second_parent] = value_to_write
